lib/modules/manager/npm/post-update/npm.ts

```diff
@@ -81,11 +83,80 @@ export async function generateLockFile(
 
     // rangeStrategy = update-lockfile
     const lockUpdates = upgrades.filter((upgrade) => upgrade.isLockfileUpdate);
-    if (lockUpdates.length) {
+
+    const lockRootUpdates: Upgrade<Record<string, any>>[] = []; // stores all upgrades which are present in root package.json
+    const lockWorkspaceUpdates: Upgrade<Record<string, any>>[] = []; // stores all upgrades which are present in workspaces package.json
+    const workspaces: Set<string> = new Set(); // name of all workspaces
+    const rootDeps: Set<string> = new Set(); // depName of all upgrades in root package.json (makes it check duplicate deps in root)
+
+    // divide the deps in two categories: workspace and root
+    for (const upgrade of lockUpdates) {
+      if (upgrade.isLockfileUpdate && upgrade.managerData?.hasWorkspaces) {
+        const workspacePatterns = upgrade.managerData?.workspacesPackages; // glob pattern or directory name/path
+        const packageFileDir = upgrade.packageFile?.replace('package.json', '');
+
+        // workspaceDir = packageFileDir - lockFileDir (root/workspace - root = workspace)
+        const workspaceDir = trimLeadingSlash(
+          packageFileDir?.replace(lockFileDir, '') ?? ''
+        );
+
+        // if packageFileDir === lockFileDir, dep is present in root package.json
+        if (packageFileDir === lockFileDir) {
+          lockRootUpdates.push(upgrade);
+          rootDeps.add(`${upgrade.depName!}@${upgrade.newVersion!}`);
```

Comment:
```suggestion
          rootDeps.add(`${upgrade.packageName}@${upgrade.newVersion!}`);
```
i think you now need to use `packageName` since v36

Reply:
Has this already been implemented, the issue tracking this is still open https://github.com/renovatebot/renovate/issues/16012